packages/itwin/tree-widget/src/components/trees/stateless/common/components/BaseTree.tsx

```diff
@@ -64,98 +79,97 @@ export function FilterableTree({ imodel, getSchemaContext, hierarchyLevelSizeLim
   getHierarchyDefinition,
   selectionMode,
   onPerformanceMeasured,
   reportUsage,
-}: Omit<FilterableTreeProps, "getSchemaContext"> & { imodelAccess: IModelAccess; defaultHierarchyLevelSizeLimit: number }) {
+  treeRenderer,
+  density,
+  getIcon,
+  getSublabel,
+}: Omit<BaseTreeProps, "getSchemaContext"> & { imodelAccess: IModelAccess; defaultHierarchyLevelSizeLimit: number }) {
   const localizedStrings = useHierarchiesLocalization();
   const {
     rootNodes,
     isLoading,
     reloadTree,
-    setFormatter: _setFormatter,
     selectNodes,
+    setFormatter: _setFormatter,
     expandNode,
     ...treeProps
   } = useUnifiedSelectionTree({
+    imodelAccess,
+    getHierarchyDefinition,
+    getFilteredPaths,
     imodelKey: imodel.key,
     sourceName: treeName,
-    imodelAccess,
     localizedStrings,
-    getHierarchyDefinition,
     onPerformanceMeasured,
     onHierarchyLimitExceeded: () => reportUsage?.({ featureId: "hierarchy-level-size-limit-hit", reportInteraction: false }),
   });
 
   useIModelChangeListener({ imodel, action: useCallback(() => reloadTree({ dataSourceChanged: true }), [reloadTree]) });
+  const reportingSelectNodes = useReportingAction({ action: selectNodes, reportUsage });
+  const { onNodeClick, onNodeKeyDown } = useSelectionHandler({ rootNodes, selectNodes: reportingSelectNodes, selectionMode: selectionMode ?? "single" });
   const { filteringDialog, onFilterClick } = useHierarchyLevelFiltering({
     imodel,
     getHierarchyLevelDetails: treeProps.getHierarchyLevelDetails,
     defaultHierarchyLevelSizeLimit,
     reportUsage,
   });
   const reportingExpandNode = useReportingAction({ action: expandNode, reportUsage });
-  const reportingSelectNodes = useReportingAction({ action: selectNodes, reportUsage });
   const reportingOnFilterClicked = useReportingAction({ action: onFilterClick, reportUsage });
 
-  const renderContent = () => {
-    if (rootNodes === undefined) {
-      return (
-        <Flex alignItems="center" justifyContent="center" flexDirection="column" style={{ width, height }}>
-          <Delayed show={true}>
-            <ProgressRadial size="large" />
-          </Delayed>
-        </Flex>
-      );
-    }
-
-    if ((rootNodes.length === 0 && !isLoading) || (rootNodes.length === 1 && !isPresentationHierarchyNode(rootNodes[0]) && rootNodes[0].type === "Unknown")) {
-      return (
-        <Flex alignItems="center" justifyContent="center" flexDirection="column" style={{ width, height }}>
-          {noDataMessage ? noDataMessage : <Text>{TreeWidget.translate("stateless.dataIsNotAvailable")}</Text>}
-        </Flex>
-      );
-    }
+  if (rootNodes === undefined) {
+    return (
+      <Flex alignItems="center" justifyContent="center" flexDirection="column" style={{ width, height }}>
+        <Delayed show={true}>
+          <ProgressRadial size="large" />
+        </Delayed>
+      </Flex>
+    );
+  }
 
+  if (rootNodes.length === 0 && !isLoading) {
```

Comment:
You removed the condition that was used to display the `noDataMessage` when hierarchy level creation fails for root nodes. It was used for `StatelessExternalSourcesTree` to show the message when `ExternalSources` class is not found in the schema. Maybe there is some better way to handle the case?

Reply:
I don't think previous approach was good. Changed `ExternalSourcesTreeDefinition` to check if schema has `ExternalSource` class